File: src/main/java/at/xirado/bean/translation/I18n.java
package at.xirado.bean.translation;

import net.dv8tion.jda.api.utils.data.DataObject;
import net.dv8tion.jda.api.utils.data.DataType;
import net.dv8tion.jda.internal.utils.Checks;

public class I18n
{
    private final String name;
    private final DataObject object;

    public I18n(String name, DataObject object)
    {
        this.name = name;
        this.object = object;
    }

    public String getString(String query, Attribute... attributes)
    {
        Checks.notNull(query, "Query");
        String[] querySplit = query.split("\\.");

        if (object.isType(querySplit[0], DataType.STRING))
            return format(object.getString(querySplit[0]), attributes);

        if (object.isType(querySplit[0], DataType.ARRAY))
            throw new IllegalArgumentException(querySplit[0] + " is an array! (" + getName() + ")");

        if (object.isNull(querySplit[0]))
            throw new NullPointerException(querySplit[0] + " is null! (" + getName() + ")");

        DataObject current = object.getObject(querySplit[0]);

        for (int i = 1; i < querySplit.length; i++)
        {
            if (current.isType(querySplit[i], DataType.OBJECT))
            {
                current = current.getObject(querySplit[i]);
                continue;
            }
            if (current.isType(querySplit[i], DataType.ARRAY))
                throw new IllegalArgumentException(querySplit[i] + " is an array! (" + getName() + ")");

            if (current.isNull(querySplit[i]))
                throw new NullPointerException(querySplit[i] + " is null! (" + getName() + ")");

            return format(current.getString(querySplit[i]), attributes);
        }
        return null;
    }

    private static String format(String pattern, Attribute... attributes)
    {
        String output = pattern;
        for (Attribute attribute : attributes)
        {
            output = output.replace("{" + attribute.getKey() + "}", attribute.getValue());
        }
        return output;
    }

    public String getName()
    {
        return name;
    }
}

File: src/main/java/at/xirado/bean/translation/I18n.kt
package at.xirado.bean.translation

import net.dv8tion.jda.api.utils.data.DataObject
import net.dv8tion.jda.api.utils.data.DataType

class I18n(val languageObject: DataObject, val name: String)
{

    fun get(query: String, vararg attributes: Attribute): String {
        val path = ArrayList(query.split("\\."))

        check(languageObject, path[0])

        if (canBeString(languageObject, path[0]))
            return format(languageObject.getString(path[0]), *attributes)

        var current = languageObject.getObject(path[0])

        for (i in 1..path.size) {
            if (current.isType(path[i], DataType.OBJECT)) {
                current = current.getObject(path[i])
                continue
            }

            check(current, path[i])

            return format(current.getString(path[i]), *attributes)
        }
        throw NullPointerException("[$name]: Key $query does not exist!")
    }

    private fun canBeString(dataObject: DataObject, key: String): Boolean {
        if (dataObject.isNull(key))
            return false

        return dataObject.isType(key, DataType.STRING)
                || dataObject.isType(key, DataType.BOOLEAN)
                || dataObject.isType(key, DataType.FLOAT)
                || dataObject.isType(key, DataType.INT)
    }

    private fun check(dataObject: DataObject, path: String) {
        if (dataObject.isNull(path))
            throw NullPointerException("[${name}]: Key $path does not exist!")

        if (languageObject.isType(path, DataType.ARRAY))
            throw IllegalArgumentException("[${name}]: Cannot access from array $path")
    }

    companion object {
        private fun format(layout: String, vararg attributes: Attribute): String {
            var output = layout

            for (attribute in attributes)
                output = output.replace("{${attribute.key}}", attribute.value)

            return output
        }
    }
}
